File: cpp/1005.maximize-sum-of-array-after-k-negations.cpp
#include <cstdlib>
#include <numeric>
#include <vector>
using std::vector;

class Solution {
public:
  int largestSumAfterKNegations(vector<int> &nums, int k) {
    // sort descending order by an absolute value
    std::sort(nums.begin(), nums.end(),
              [&](int a, int b) { return abs(a) > abs(b); });

    // use k to make biggest negative values positive
    for (int i = 0; i < nums.size() && k > 0; ++i) {
      if (nums[i] < 0) {
        nums[i] = -nums[i];
        --k;
      }
    }

    // if it's still k left, just make
    // smallest absolute value negative
    if (k % 2) {
      nums.back() = -nums.back();
    }

    // return total sum
    return std::accumulate(nums.begin(), nums.end(), 0);
  }
};

File: python/1005.maximize-sum-of-array-after-k-negations.py
class Solution:
    def largestSumAfterKNegations(self, nums: list[int], k: int) -> int:

        nums.sort(key=abs, reverse=True)
        n = len(nums)

        for i in range(n):
            if nums[i] < 0:
                nums[i] = -nums[i]
                k -= 1

            if not k:
                break

        if k % 2:
            nums[-1] = -nums[-1]

        return sum(nums)

